### PR TITLE
RavenDB-12376: librvnpal

### PR DESCRIPTION
**Pull request by aviviadi**
currently implements WriteFileHeader in linux-x64 (tested)
TODO: other plats&archs plus implement publish copy of libs/librvnpal/* (csproj)

### CODE REVIEW

File: src/Voron/Platform/Pal.cs [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+// ReSharper disable InconsistentNaming
+
+namespace Voron.Platform
+{
+    public static unsafe class Pal
+    {
+        static Pal()
+        {
+            var toFilename = LIBRVNPAL;
+            var fromFilename = $"{toFilename}.win.x64.dll";
```
Comment (ayende) [inline src/Voron/Platform/Pal.cs]: why is this initialized here?

File: src/Voron/Platform/Pal.cs [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+            {
+                fromFilename = Environment.Is64BitProcess ? $"{toFilename}.mac.x64.dylib" : $"{toFilename}.mac.x86.dylib";
+                toFilename += ".dylib";
+            }
+            else if (RuntimeInformation.IsOSPlatform(OSPlatform.Windows))
+            {
+                fromFilename = Environment.Is64BitProcess ? $"{toFilename}.win.x64.dll" : $"{toFilename}.win.x86.dll";
+                toFilename += ".dll";
+            }
+            else
```
Comment (ayende) [inline src/Voron/Platform/Pal.cs:43]: ARM missing

File: src/Voron/Platform/PalFlags.cs [hunk 3]
```diff
@@ -0,0 +1,189 @@
...
+// ReSharper disable InconsistentNaming
+// ReSharper disable UnusedMember.Local
+// ReSharper disable IdentifierTypo
+
+namespace Voron.Platform
+{
+    public static class PalFlags
+    {
+        [Flags]
+        public enum Errno
```
Comment (ayende) [inline src/Voron/Platform/PalFlags.cs]: Let's remove this.
I would rather have a dedicated function in PAL that I can hand an error code and get a string back that would hold the error string.

File: src/Voron/Platform/PalHelper.cs [hunk 4]
```diff
@@ -0,0 +1,25 @@
+using System;
+using System.IO;
+using static Voron.Platform.PalFlags;
+// ReSharper disable StringLiteralTypo
+
+namespace Voron.Platform
+{
+    public static class PalHelper
+    {
+        public static void ThrowLastError(PalFlags.Errno lastError, string msg)
```
Comment (ayende) [inline src/Voron/Platform/PalHelper.cs]: Call to PAL, which will allocate a string, which we'll copy and then free.

File: src/Voron/StorageEnvironmentOptions.cs [hunk 5]
```diff
@@ -666,10 +667,10 @@ public override unsafe bool ReadHeader(string filename, FileHeader* header)
             public override unsafe void WriteHeader(string filename, FileHeader* header)
             {
                 var path = _basePath.Combine(filename);
-                if (RunningOnPosix)
-                    PosixHelper.WriteFileHeader(header, path);
-                else
-                    Win32Helper.WriteFileHeader(header, path);
+                var errorCode = PalFlags.Errno.NONE;
+                var rc = Pal.rvn_write_header(path.FullPath, header, (uint)sizeof(FileHeader), ref errorCode);
```
Comment (ayende) [inline src/Voron/StorageEnvironmentOptions.cs]: You can define this as `out`, not `ref`

File: src/librvnpal/inc/rvn.h [hunk 6]
```diff
@@ -0,0 +1,23 @@
+#ifndef RVN_H
+#define RVN_H
+
+#include <stdint.h>
+
+#if _WIN32
+_declspec(dllexport)
+#endif
+int32_t rvn_memcmp(const void *s1, const void *s2, int32_t n);
```
Comment (ayende) [inline src/librvnpal/inc/rvn.h]: Remove this

File: src/librvnpal/inc/rvn.h [hunk 7]
```diff
@@ -0,0 +1,23 @@
...
+
+#if _WIN32
+_declspec(dllexport)
+#endif
+uint64_t rvn_get_current_thread_id(void);
+
+#if _WIN32
+_declspec(dllexport)
+#endif
+int32_t rvn_write_header(const char *path, void *header, int32_t size, uint32_t *detailed_error_code);
```
Comment (ayende) [inline src/librvnpal/inc/rvn.h]: `size` is defined as `int32_t` here, but `uint` in the pinvoke sig

File: src/librvnpal/src/posix/writefileheader.c [hunk 8]
```diff
@@ -0,0 +1,248 @@
...
+
+#ifndef CIFS_MAGIC_NUMBER
+	#define CIFS_MAGIC_NUMBER     0xff534d42
+#endif
+
+typedef int bool;
+#define true 1
+#define false 0
+
+#define SUCCESS 			0
```
Comment (ayende) [inline src/librvnpal/src/posix/writefileheader.c]: Shouldn't this be moved to header file?

File: src/librvnpal/src/posix/writefileheader.c [hunk 9]
```diff
@@ -0,0 +1,248 @@
...
+	int rc;
+	int fd = open(dir_path, 0, 0);
+    if (fd == -1)
+    {
+		rc = FAIL_OPEN_FILE;
+		goto error_cleanup;
+	}
+
+	rc = sync_directory_allowed(fd);
+	if(rc == SYNC_DIR_ALLOWED ||
```
Comment (ayende) [inline src/librvnpal/src/posix/writefileheader.c]: If checking if synced failed, we always sync ?

File: src/librvnpal/src/posix/writefileheader.c [hunk 10]
```diff
@@ -0,0 +1,248 @@
...
+
+error_cleanup:
+	*detailed_error_code = errno;
+	if(fd != -1)
+		close(fd);
+	
+	return rc;
+}
+
+static int32_t sync_directory_maybe_symblink(char *dir_path, int depth, uint32_t* detailed_error_code)
```
Comment (ayende) [inline src/librvnpal/src/posix/writefileheader.c]: The whole code need re-formatting

**Review by ayende — CHANGES_REQUESTED** [review]

File: src/Voron/Platform/Pal.cs (hunk 1 above)
Comment (aviviadi) [reply src/Voron/Platform/Pal.cs]: it works (because executed before lib load) and makes logic to put it here in terms of context

Comment (ayende) [reply src/Voron/Platform/Pal.cs]: Why is it init to `$"{toFilename}.win.x64.dll";` ?

**Pull request merged**